File: backend/routers/images.php
<?php

$image_query = require $_SERVER["DOCUMENT_ROOT"] . "/queries/images.php";
$request_path = parse_url($_SERVER['REQUEST_URI'], PHP_URL_PATH);

if ($_SERVER["REQUEST_METHOD"] == "GET") {
    switch ($request_path) {
        case (str_contains($request_path, "/GetAllImages/")):
            $images = $image_query->GetAllImages();

            $images = $image_query->OrderImageList($images);

            $encoded_images = [];
            foreach ($images as $image) {
                array_push($encoded_images, $image->ToObject());
            }
            echo json_encode($encoded_images);
            return json_encode($encoded_images);


        case (str_contains($request_path, "/GetSomeImages")):
            $searchQuery = explode("?", $_SERVER['REQUEST_URI']);
            $data = explode("&", $searchQuery[1]);

            $searchQueries = [];
            foreach ($data as $individualSQ){
                $individualSQ = explode("=", $individualSQ);
                $searchQueries[$individualSQ[0]] = "%" . $individualSQ[1] . "%";
            }

            if ($searchQueries["mediums"] == "%%" && $searchQueries["search"] != "%%") {
                $searchQueries["mediums"] = $searchQueries["search"];
            }
            if ($searchQueries["tags"] == "%%" && $searchQueries["search"] != "%%") {
                $searchQueries["tags"] = $searchQueries["search"];
            }
            
            
            $images = $image_query->SearchAllImages($searchQueries);
            $images = $image_query->OrderImageList($images);
            $encoded_images = [];
            foreach ($images as $image) {
                array_push($encoded_images, $image->ToObject());
            }

            echo json_encode($encoded_images);
            return json_encode($encoded_images);

        
        case (str_contains($request_path, "/GetImage/")):
            break;
    }
}

if ($_SERVER["REQUEST_METHOD"] == "POST") {
    switch ($request_path) {
        case (str_contains($request_path, "/Upload/")):
            $image_data = json_decode(file_get_contents("php://input"));
            
            $data = explode( ',', $image_data->image_base64 );

            $fileType = explode(";", $data[0]);
            $fileType = explode("/", $fileType[0])[1];

            $saveLocal = getenv('Image_Repo');
            $filename = strval(count(glob($saveLocal . "*.{jpeg,jpg,png}",GLOB_BRACE)) + 1) . ".$fileType";

            $image_file = fopen($filename, "w");
            fwrite( $image_file, base64_decode( $data[1] ) );
            fclose( $image_file ); 
            rename($filename, $saveLocal . $filename);

            $image = new Image($filename, $image_data->title, $image_data->tags, $image_data->mediums, $image_data->created);
            $res = $image_query->PostImage($image);
            echo $res;
            return $res;
    }
}

File: backend/routers/images.py
import os

from typing import Annotated
from PIL import Image as IMG

from fastapi import (APIRouter, Depends, Header)
from fastapi.responses import FileResponse

from queries.images import ImageQuery, OrderStyle
from models.images import *

router = APIRouter(tags=["IMAGES"])

@router.get("/api/images/{image_id}")
async def get_image_file(
    image_id: int,
    file: str | None = None,
    compressed: bool | None = None,
    repo: ImageQuery = Depends()
):
    image = repo.get_one_image(image_id)
    if file:
        image_file = repo.get_image_file(image)

        if compressed:
            compressed_image = IMG.open(image_file)
            compressed_image.save(
                image_file,
                "JPEG",
                optimize=True,
                quality = 60
            )
            return FileResponse(image_file)
        return FileResponse(image_file)
    return image


@router.get("/api/images")
async def get_all_images(
        title: str | None =  None,
        mediums: str | None = None,
        tags: str | None = None,
        created: str | None = None,
        sort_style: OrderStyle | None = OrderStyle.DEFAULT,
        range_start: int | None = None,
        range_end: int | None = None,
        repo: ImageQuery = Depends()
    ):

    if range_start or range_end:
        images = repo.get_range_of_images(range_start, range_end)
        return repo.order_image_list(images, sort_style)
    
    if title or mediums or tags or created:
        if title == None: title = ""
        if created == None: created = ""
        if mediums == None: mediums = ""
        if tags == None: tags = ""
        
        search = {
            "title": title,
            "mediums": mediums,
            "tags": tags,
            "created": created
        }
        images = repo.get_some_images(search)
        return repo.order_image_list(images, sort_style)
    
    images = repo.get_all_images()

    if (sort_style):
        return repo.order_image_list(images, sort_style)
    else:
        return images


@router.post("/api/images/")
async def post_image(
    image: ImageIn,
    api_access_token: str = Header(..., convert_underscores=True),
    repo: ImageQuery = Depends()
):
    if (not api_access_token == os.environ.get("GALLERY_API_ACCESS_TOKEN")): return {"Error": "Invalid api token"}
    filename = repo.upload_image_file(image)
    if filename:
        return repo.post_image(image, filename)
    else:
        return {"Error": "Image failed to post."}